kem/mlkem/acvp_test.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return nil, err
+	}
+	return gunzip(buf)
+}
+
+func TestACVP(t *testing.T) {
+	for _, sub := range []string{
+		"keyGen",
+		"encapDecap",
+	} {
+		t.Run(sub, func(t *testing.T) {
+			testACVP(t, sub)
+		})
+	}
+}
+
+// nolint:funlen,gocyclo
+func testACVP(t *testing.T, sub string) {
+	buf, err := readGzip("testdata/ML-KEM-" + sub + "-FIPS203/prompt.json.gz")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var prompt struct {
+		TestGroups []json.RawMessage `json:"testGroups"`
+	}
+
+	if err = json.Unmarshal(buf, &prompt); err != nil {
+		t.Fatal(err)
+	}
+
+	buf, err = readGzip("testdata/ML-KEM-" + sub + "-FIPS203/expectedResults.json.gz")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var results struct {
+		TestGroups []json.RawMessage `json:"testGroups"`
+	}
+
+	if err := json.Unmarshal(buf, &results); err != nil {
+		t.Fatal(err)
+	}
+
+	rawResults := make(map[int]json.RawMessage)
+
+	for _, rawGroup := range results.TestGroups {
+		var abstractGroup struct {
+			Tests []json.RawMessage `json:"tests"`
+		}
+		if err := json.Unmarshal(rawGroup, &abstractGroup); err != nil {
+			t.Fatal(err)
+		}
+		for _, rawTest := range abstractGroup.Tests {
+			var abstractTest struct {
+				TcID int `json:"tcId"`
+			}
+			if err := json.Unmarshal(rawTest, &abstractTest); err != nil {
+				t.Fatal(err)
+			}
+			if _, exists := rawResults[abstractTest.TcID]; exists {
+				t.Fatalf("Duplicate test id: %d", abstractTest.TcID)
+			}
+			rawResults[abstractTest.TcID] = rawTest
+		}
+	}
+
+	for _, rawGroup := range prompt.TestGroups {
+		var abstractGroup struct {
+			TestType string `json:"testType"`
+		}
+		if err := json.Unmarshal(rawGroup, &abstractGroup); err != nil {
+			t.Fatal(err)
+		}
+		switch {
+		case abstractGroup.TestType == "AFT" && sub == "keyGen":
+			var group struct {
+				TgID         int    `json:"tgId"`
+				ParameterSet string `json:"parameterSet"`
+				Tests        []struct {
+					TcID int      `json:"tcId"`
+					Z    HexBytes `json:"z"`
+					D    HexBytes `json:"d"`
+				}
+			}
+			if err := json.Unmarshal(rawGroup, &group); err != nil {
+				t.Fatal(err)
+			}
+
+			scheme := schemes.ByName(group.ParameterSet)
+			if scheme == nil {
+				t.Fatalf("No such scheme: %s", group.ParameterSet)
+			}
+
+			for _, test := range group.Tests {
+				var result struct {
+					Ek HexBytes `json:"ek"`
+					Dk HexBytes `json:"dk"`
+				}
+				rawResult, ok := rawResults[test.TcID]
+				if !ok {
+					t.Fatalf("Missing result: %d", test.TcID)
+				}
+				if err := json.Unmarshal(rawResult, &result); err != nil {
+					t.Fatal(err)
+				}
+
+				var seed [64]byte
+				copy(seed[:], test.D)
+				copy(seed[32:], test.Z)
+
+				ek, dk := scheme.DeriveKeyPair(seed[:])
+
+				ek2, err := scheme.UnmarshalBinaryPublicKey(result.Ek)
+				if err != nil {
+					t.Fatalf("tc=%d: %v", test.TcID, err)
+				}
+				dk2, err := scheme.UnmarshalBinaryPrivateKey(result.Dk)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if dk.Equal(dk2) {
```

Comment:
Based on the error message, aren't these two conditions inverted?
```suggestion
				if !dk.Equal(dk2) {
```

Reply:
Oof. Well spotted. Let me figure out why it fails...

Reply:
Ok, apparently the ACVP test vectors [haven't been updated yet](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/s_Wez9FanHw/m/YGB-zccRBgAJ).